integrations/near/include/marlin/near/OnRampNear.hpp

```diff
@@ -0,0 +1,203 @@
+		);
+		if(b58enc(b58, &sz, clop.static_pk, 32)) {
+			SPDLOG_INFO(
+				"{}",
+				b58
+			);
+		} else {
+			printf("Failed\n");
+		}
+
+		f.bind(SocketAddress::loopback_ipv4(8000));
+		f.listen(*this);
+	}
+
+	void did_recv_message(NearTransport <OnRampNear> &, Buffer &&message) {
+		SPDLOG_INFO(
+			"msgSize: {}; Message received from NearTransport: {}",
+			message.size(),
+			spdlog::to_hex(message.data(), message.data() + message.size())
+		);
+		if(message.data()[0] == 0x10) {
+			handle_handshake(std::move(message));
+		} else if(message.data()[0] == 0xc) {
+			handle_transaction(std::move(message));
+		} else if(message.data()[0] == 0xb) {
+			handle_block(std::move(message));
+		} else if(message.data()[0] == 0xd) {
+			SPDLOG_INFO(
+				"This is a RoutedMessage"
+			);
+		} else {
+			SPDLOG_INFO(
+				"Something else: {}",
+				message.data()[0]
+			);
+		}
+	}
+
+	template<typename T> // TODO: Code smell, remove later
+	void did_recv_message(
+		DefaultMulticastClient<OnRampNear> &,
+		Buffer &&,
+		T,
+		uint16_t,
+		uint64_t
+	) {}
+
+	void did_send_message(NearTransport<OnRampNear> &, Buffer &&message) {
+		SPDLOG_INFO(
+			"Transport: Did send message: {} bytes",
+			// transport.src_addr.to_string(),
+			// transport.dst_addr.to_string(),
+			message.size()
+		);
+	}
+
+	void did_dial(NearTransport<OnRampNear> &) {
+	}
+
+	bool should_accept(SocketAddress const &) {
+		return true;
+	}
+
+
+	void did_subscribe(
+		DefaultMulticastClient<OnRampNear> &,
+		uint16_t
+	) {}
+
+	void did_unsubscribe(
+		DefaultMulticastClient<OnRampNear> &,
+		uint16_t
+	) {}
+
+	void did_close(NearTransport<OnRampNear> &, uint16_t) {}
+
+	void did_create_transport(NearTransport <OnRampNear> &transport) {
+		nearTransport = &transport;
+		transport.setup(this);
+	}
+};
+
+
+void OnRampNear::handle_transaction(core::Buffer &&message) {
+	SPDLOG_DEBUG(
+		"Handling transaction: {}",
+		spdlog::to_hex(message.data(), message.data() + message.size())
+	);
+
+	CryptoPP::BLAKE2b blake2b((uint)8);
+	blake2b.Update((uint8_t *)message.data(), message.size());
+	uint64_t message_id;
+	blake2b.TruncatedFinal((uint8_t *)&message_id, 8);
+	multicastClient.ps.send_message_on_channel(
+		1,
+		message_id,
+		message.data(),
+		message.size()
+	);
+}
+
+void OnRampNear::handle_block(core::Buffer &&message) {
+	SPDLOG_INFO("Handling block");
+
+	CryptoPP::BLAKE2b blake2b((uint)8);
+	blake2b.Update((uint8_t *)message.data(), message.size());
+	uint64_t message_id;
+	blake2b.TruncatedFinal((uint8_t *)&message_id, 8);
+	multicastClient.ps.send_message_on_channel(
+		0,
+		message_id,
+		message.data(),
+		message.size()
+	);
+}
+
+void OnRampNear::handle_handshake(core::Buffer &&message) {
+	SPDLOG_INFO("Replying");
+	uint8_t *buf = message.data();
+	uint32_t buf_size = message.size();
+	std::swap_ranges(buf + 9, buf + 42, buf + 42);
+
+	uint8_t near_key_offset = 42, gateway_key_offset = 9;
+
+	uint8_t msg_sig[74];
```

Comment:
Are you using this outside of the hashing? If not, you can directly build up the hash in pieces rather than constructing this.